dict_autotrainer.cc

```diff
@@ -0,0 +1,160 @@
+        auto training_sample = co_await _ss.do_sample_sstables(s->id(), params.chunk_length(), n_chunks);
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: got training sample with {} chunks", s->ks_name(), s->cf_name(), training_sample.size());
+        if (training_sample.size() < n_chunks) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: not enough chunks, giving up on the update", s->ks_name(), s->cf_name());
+            continue;
+        }
+        auto dict = co_await _ss.train_dict(std::move(training_sample));
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: trained dict of size {}", s->ks_name(), s->cf_name(), dict.size());
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: sampling for validation with chunk_length={}, n_chunks={}", s->ks_name(), s->cf_name(), params.chunk_length(), n_chunks);
+        auto validation_sample = co_await _ss.do_sample_sstables(s->id(), params.chunk_length(), n_chunks);
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: got validation sample with {} chunks", s->ks_name(), s->cf_name(), validation_sample.size());
+        if (validation_sample.size() < n_chunks) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: not enough chunks, giving up on the update", s->ks_name(), s->cf_name());
+            continue;
+        }
+        auto ratio_before = co_await try_one_compression_config(_ss.get_database().get_user_sstables_manager().get_compressor_factory(), s, params, validation_sample);
+        auto ratio_after = co_await try_one_compression_config(dict, s, params, validation_sample);
+        constexpr float threshold_fraction = 0.95;
+        float threshold = threshold_fraction * ratio_before;
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: ratio_before={}, ratio_after={}, update_threshold={}", s->ks_name(), s->cf_name(), ratio_before, ratio_after, threshold);
+        if (ratio_after > threshold) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: ratio_after below threshold. Discarding the new dict.", s->ks_name(), s->cf_name());
+            continue;
+        }
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: Publishing the new dict.", s->ks_name(), s->cf_name());
+        co_await _ss.publish_new_sstable_dict(s->id(), dict, _group0_client);
+    }
+}
+
+future<> sstable_dict_autotrainer::run() {
+    SCYLLA_ASSERT(this_shard_id() == 0);
+    alogger.debug("sstable_dict_autotrainer::run(): starting");
+    std::default_random_engine rng(0);
+    while (!_as.abort_requested()) {
+        try {
+            auto tick_period = std::chrono::duration<float>(_cfg.tick_period_in_seconds());
+            auto tick_period_ms = std::chrono::duration_cast<std::chrono::milliseconds>(tick_period);
+            co_await seastar::sleep_abortable(tick_period_ms, _as);
+            if (_ss.is_raft_leader() && _ss.get_feature_service().sstable_compression_dicts) {
```

Comment:
I don't think this is reliable. Someone might think they are the leader but only discover they're not after trying to commit something. What happens if two leaders clash?

I'm not sure if this is the right way to integrate. Maybe better is to hook it into the topology change coordinator. @patjed41 for opinions on the proper raft way.

Reply:
The leader check is just there to avoid noise. Everything should work correctly even if every node is publishing dicts, but it just may cause more dicts to be flying around than necessary.

The dict is just a recommendation. If two leaders clash, in the worst case some node will use a "short-lived" dict (which is immediately superseded by another one) for an sstable, and now it will have to keep two dicts in memory instead of one, while that sstable is alive.

Reply:
The "right way" would be to have some sort of a distributed cron job. Which, I guess, topology coordinator could be, but we are already stuffing too much topology-unrelated logic into the critical Raft places.

Reply:
Using *group0* for this is icky in the first place. Non-critical things like this should probably be in their own group.